### DISCUSSION TITLE
ANIMATIONS ADD PROPS

### DISCUSSION THREAD

**Discussion by ghndy2** [YimMenu/YimMenu#2130] [Q&A]
did anyone know how to add some props when us doing animation like fivem script did?

**Comment by pierrelasse** [comment]
???

**Comment by Rabtos** [comment]
Its called a scenario if you add props. I want it too...

**Comment by Tessio** [comment]
+1

**Comment by lonelybud** [comment]
to play scenario -

https://docs.fivem.net/natives/?_0x142A02425FF02BD9

https://github.com/DurtyFree/gta-v-data-dumps/blob/master/scenariosCompact.json

```
TASK::TASK_START_SCENARIO_IN_PLACE(ped, scenarioName, duration_ms, 0);
```

set duration_ms to -1 for infinite duration

use `TASK::CLEAR_PED_TASKS(ped)` to stop

talking about original answer. adding props to some animation. This is something not present right now.

**Reply by xesdoog** [reply]
Hello @lonelybud, how have you been?
Sorry to bother you with this but you've been a helpful member of the community so if you don't mind, I need your insight on an issue related to this discussion whenever you find the time to reply.
I made an [animation script](https://github.com/YimMenu-Lua/SAMURAI-Animations) that does exactly what is mentioned above but after playing a bunch of anims, it starts causing major FPS drops. I did some troubleshooting to find out why this was happening and it turns out once you request an animation dictionary, it will stay loaded (or keep requesting it over and over? idk) so playing several animations will load a bunch of anim dicts without ever getting rid of them after you're done with them therefore causing the game to start running slow.
The steps I've gone through to make sure it actually is the anim dict causing this and not something else :
- Only play animations that do not request any props or particle effects.
- Add a 5s delay at the bottom of the **Stop button** and the **hotkey** _(both are supposed to unload everything and stop the animation)_ then check to see what was still loaded? model hashes? ptfx? anim dicts? then toast a message telling me what was still there. I got a positive on the anim dict.
- Lastly, play a few basic animations then go to the script code and quote out the part that requests the anim dict and anim set, reload the script and test. The animations that were played before will play again no problem but the ones that weren't played will not do anything meaning the previously requested dicts were still loaded.

I added ```STREAMING.REMOVE_ANIM_DICT()``` in every part I can think of to no avail. I'm lost for for solutions because I'm not skilled in any way I'm still learning as I go but this is starting to frustrate me. I hate it when I make a mess and can't fix it.

**Reply by lonelybud** [reply]
the comment i added are for scenarios and not animations. In your Lua script you are using animations and while doing it you add props as well to simulate a kind of scenario. So it is different completely. You should properly load anim dict, ptfx dict etc do unload them after use for memory related issues.

I dont see how it can effect FPS perhaps something else must be going on?

example -
```c++
	STREAMING::REQUEST_NAMED_PTFX_ASSET(...);
	GRAPHICS::USE_PARTICLE_FX_ASSET(...);
	GRAPHICS::START_NETWORKED_PARTICLE_FX_NON_LOOPED_AT_COORD(...);

         // after you are done or switching to another scenario
         STREAMING::REMOVE_NAMED_PTFX_ASSET(...);
         // or
         STREAMING::REMOVE_PTFX_ASSET(); 
```
For animations you can check - https://github.com/YimMenu/YimMenu/blob/9855e46dd164df6c00d5f6b212ef5084e68c7785/src/util/ped.hpp#L563
then you can unload it as
`STREAMING::REMOVE_ANIM_DICT(...);`

you can also check for if dict exist or not using DOES_ANIM_DICT_EXIST. Similar checks for ptfx can be added `HAS_NAMED_PTFX_ASSET_LOADED` etc, 

Similarly you need to spawn objects from yims lua api - CREATE_OBJECT and delete them when playing other scenario etc.

If you look at the original code in Yim menu itself they are not unloading the anim dict after use and still it works fine. But for your case you can add as it is good to get rid of resources.

Try to rewrite your code and try to understand it from scratch. May other devs can also help you

**Reply by lonelybud** [reply]
@Yimura sorry to interrupt you but I want to ask if is lua api OBJECT.CREATE_OBJECT safe I mean does it include object bypass?

**Reply by xiaoxiao921** [reply]
it does not, also I don't think they ban for this, the object simply won't spawn if it's blacklisted

**Reply by xesdoog** [reply]
I'll try to rewrite it from scratch because I got ahead of myself and started implementing animations that have ptfx and more than one prop like cooking on a bbq and such so I made a huge mess lol
Thanks for replying though, I really appreciate it
